jormungandr/src/fragment/pool.rs

```diff
@@ -85,12 +65,20 @@ impl Pool {
                         .zip(fragments_exist_in_logs)
                         .filter(|(_, exists_in_logs)| !exists_in_logs)
                         .map(|(fragment, _)| fragment);
-                    let new_fragment_ids = pool.insert_all(new_fragments);
-                    let count = new_fragment_ids.len();
-                    let fragment_logs = new_fragment_ids
-                        .into_iter()
-                        .map(move |id| FragmentLog::new(id.into(), origin));
-                    logs.insert_all(fragment_logs).map(move |_| count)
+                    let new_fragments = pool.insert_all(new_fragments);
+                    let count = new_fragments.len();
+                    let fragment_logs = new_fragments
+                        .iter()
+                        .map(move |fragment| FragmentLog::new(fragment.id().into(), origin))
+                        .collect::<Vec<_>>();
+                    stream::iter_ok(new_fragments)
+                        .map(|fragment| NetworkMsg::Propagate(PropagateMsg::Fragment(fragment)))
+                        .for_each(move |network_msg| network_msg_box.try_send(network_msg))
```

Comment:
More async, but still `try_send`? It's not recommended by the `Sender` docs. You can make it overflow-proof by doing `send` (of trait `Sink`) inside the `for_each` closure, if I'm not mistaken. Or, better and shorter, feed the whole stream of `PropagateMsg` into the message box with `send_all`.

Reply:
Sorry, a typo, I wanted to use `send`. `send_all` looks better

Reply:
`send_all` and `forward` are the only way to send whole stream at once into a sink, but both of them close the sink after flushing. There's no workaround for that other than writing custom stream consumer that pushes items one by one: https://github.com/rust-lang-nursery/futures-rs/issues/462

We can just ignore this fact, because luckily Tokio MPSC does nothing on close. This is ugly, illegal and relies on undocumented behavior. I guess that we can use that temporarily. If MPSC behavior changes before we migrate to std futures infrastructure (which is not even there), the tests will probably notice that.

Reply:
No, better just do `send` inside `for_each`.

Reply:
An incidental observation is that `SendAll` does not have the poll-after-done bug that `Forward` has in futures 0.1.29, being otherwise equivalent. I can remove the fuse workaround this way.